File: src/pratice/backtraking/N과M15663.java
package pratice.backtraking;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;


public class N과M15663 {
    public static int[] selected;
    public static int[] num, used;
    public static int N, M;
    public static StringBuffer buffer = new StringBuffer();
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String[] split = reader.readLine().split(" ");
        N = Integer.parseInt(split[0]);
        M = Integer.parseInt(split[1]);
        num = new int[N + 1];
        used = new int[N + 1];
        selected = new int[M];
        split = reader.readLine().split(" ");
        for (int i = 0; i < split.length; i++) {
            num[i + 1] = Integer.parseInt(split[i]);
        }
        Arrays.sort(num);
        function(0);
        System.out.println(buffer);
    }

    public static void function(int curIdx){
        if(curIdx >= M){
            for (int i = 0; i < curIdx; i++) {
                buffer.append(selected[i] + " ");
            }
            buffer.append("\n");
        }else{
            int lastNum = num[0];
            for (int i = 1; i <=N ; i++) {
                if( used[i] == 0 && lastNum != num[i]) {
                    lastNum = num[i];
                    selected[curIdx] = num[i]; used[i] = 1;
                    function(curIdx + 1);
                    selected[curIdx] = 0; used[i] = 0;
                }
            }
        }
    }
}

File: src/pratice/backtraking/N과M15663.kt
package pratice.backtraking

private var N = 0
private var M = 0
private var num = emptyList<Int>()
private val buffer = StringBuffer()
private var selected = IntArray(0)
private var used = BooleanArray(0)
fun main(){
    input()
    recFunc(0)
    println(buffer)
}

private fun input() {
    val ints = readln().split(" ").map { n -> n.toInt() }
    N = ints[0]
    M = ints[1]
    selected = IntArray(M)
    num = readln().split(" ").map { s -> s.toInt() }.toList()
    num = num.sorted()
    used = BooleanArray(N)
}

private fun recFunc(k: Int) {
    if(k >= M){
        for(i in 0 until  M){
            buffer.append("${selected[i]} ")
        }
        buffer.append("\n")
    }else{
        var lastNum = -1
        for(i in  0 until num.size){
           if(!used[i] && lastNum!= num[i]){
               lastNum = num[i]
               selected[k] = num[i]
               used[i] = true
               recFunc(k + 1)
               used[i] = false
               selected[k] = 0
           }
        }
    }
}
